server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -76,32 +81,74 @@ long[] buildAndWritePostingsLists(
+                docIds = new int[size];
+                clusterOrds = new int[size];
+            }
+            if (spillDocIds == null || spillDocIds.length < overspillCluster.length) {
+                spillDocIds = new int[overspillCluster.length];
+                spillClusterOrds = new int[overspillCluster.length];
+            }
+            for (int j = 0; j < size; j++) {
+                docIds[j] = floatVectorValues.ordToDoc(cluster[j]);
+                clusterOrds[j] = j;
+            }
+            for (int j = 0; j < overspillCluster.length; j++) {
+                spillDocIds[j] = floatVectorValues.ordToDoc(overspillCluster[j]);
+                spillClusterOrds[j] = j;
+            }
+            final int[] finalDocs = docIds;
+            final int[] finalOrds = clusterOrds;
+            final int[] finalSpillDocs = spillDocIds;
+            final int[] finalSpillOrds = spillClusterOrds;
+            // sort cluster.buffer by docIds values, this way cluster ordinals are sorted by docIds
+            new IntSorter(clusterOrds, i -> finalDocs[i]).sort(0, size);
+            // sort overspillCluster.buffer by docIds values, this way overspill ordinals are sorted by docIds
+            new IntSorter(spillClusterOrds, i -> finalSpillDocs[i]).sort(0, overspillCluster.length);
+            // encode doc deltas
+            for (int j = 0; j < size; j++) {
+                docIds[j] = j == 0 ? finalDocs[finalOrds[j]] : finalDocs[finalOrds[j]] - finalDocs[finalOrds[j - 1]];
+            }
+            for (int j = 0; j < overspillCluster.length; j++) {
+                spillDocIds[j] = j == 0
+                    ? finalSpillDocs[finalSpillOrds[j]]
+                    : finalSpillDocs[finalSpillOrds[j]] - finalSpillDocs[finalSpillOrds[j - 1]];
+            }
+            postingsOutput.writeInt(size);
+            postingsOutput.writeInt(overspillCluster.length);
             postingsOutput.writeInt(Float.floatToIntBits(VectorUtil.dotProduct(centroid, centroid)));
-            onHeapQuantizedVectors.reset(centroid, size, ord -> cluster[ord]);
             // TODO we might want to consider putting the docIds in a separate file
             // to aid with only having to fetch vectors from slower storage when they are required
             // keeping them in the same file indicates we pull the entire file into cache
-            docIdsWriter.writeDocIds(j -> floatVectorValues.ordToDoc(cluster[j]), size, postingsOutput);
+            postingsOutput.writeGroupVInts(docIds, size);
+            postingsOutput.writeGroupVInts(spillDocIds, overspillCluster.length);
+            onHeapQuantizedVectors.reset(centroid, size, j -> cluster[finalOrds[j]]);
+            bulkWriter.writeVectors(onHeapQuantizedVectors);
+            // write overspill vectors
+            onHeapQuantizedVectors.reset(centroid, overspillCluster.length, j -> overspillCluster[finalSpillOrds[j]]);
             bulkWriter.writeVectors(onHeapQuantizedVectors);
```

Comment:
This sort of what you had in mind @iverase ?

Reply:
yes. My idea is that we can use this information to first score the assignments of all the clusters we want to visit  so we can ensure that the posting lists will be unique and have simpler (and faster) visiting logic, and later visit the spill assignments where we would have  a more complex (slower) logic to remove already visited posting lists. The downside of this approach is that it will require more hops on the posting list files breaking a bit the disk friendly approach of this type of index.

Do you think this is something doable? it complicated the search logic quite a bit and maybe the benefits are too small. What do you think?

Reply:
So, we already have all the vectors are already in doc order. But, when we combine initial and secondary assignments into one grouping, that is when you get a partial ordering. 

Each assignment array is already in vector ordinal order, which also means that they are already in doc Id order. 

This PR now just keeps them separate (no sorting required).

Overall, there don't seem to be significant performance gains.

I noticed slightly lower performance when no filters are provided. 

I noticed higher performance with very restrictive filters are used with high-nprobe, but I don't really know why. 

I wouldn't expect doc ID decoding to be a significant issue. 

I am gonna leave this as it is and we can revisit it at a later time, unless you have a better intuition around this.